src/box/txn.c

```diff
@@ -885,6 +895,19 @@ txn_journal_entry_new(struct txn *txn)
 
 		req->approx_len += xrow_approx_len(stmt->row);
 	}
+
+	/*
+	 * Look at the flags stmt->row->flags. If the transaction
+	 * is synchronous, then set is_sync = true. This should
+	 * only be done on replicas. The master sets these flags
+	 * and independently decides whether the transaction is
+	 * synchronous or not. All txn meta flags are set only for
+	 * the last txn row.
+	 */
+	stmt = stailq_last_entry(&txn->stmts, typeof(*stmt), next);
+	if ((stmt->row->flags & IPROTO_FLAG_WAIT_ACK) != 0)
+		is_sync = true;
+
```

Comment:
AFAIU, this is needed only for remote transaction coming from the applier. Then it has to be moved into the applier code. To `apply_plain_tx()` and wherever else it starts transactions. Same for the recovery stage for transactions obtained from the xlogs. Like I said, this code (`txn_journal_entry_new()`) is very hot, it has to have as few `if`s as possible.

Reply:
Moved to applier.